optuna/storages/_rdb/storage.py

```diff
@@ -733,14 +737,36 @@ def _set_trial_user_attr_without_commit(
         trial = models.TrialModel.find_or_raise_by_id(trial_id, session)
         self.check_trial_is_updatable(trial_id, trial.state)
 
-        attribute = models.TrialUserAttributeModel.find_by_trial_and_key(trial, key, session)
-        if attribute is None:
-            attribute = models.TrialUserAttributeModel(
-                trial_id=trial_id, key=key, value_json=json.dumps(value)
+        if self.engine.name == "mysql":
+            mysql_insert_stmt = sqlalchemy_dialects_mysql.insert(
+                models.TrialUserAttributeModel
+            ).values(trial_id=trial_id, key=key, value_json=json.dumps(value))
+            mysql_upsert_stmt = mysql_insert_stmt.on_duplicate_key_update(
+                value_json=mysql_insert_stmt.inserted.value_json
             )
-            session.add(attribute)
+            session.execute(mysql_upsert_stmt)
+        elif self.engine.name == "sqlite":
+            sqlite_insert_stmt = sqlalchemy_dialects_sqlite.insert(
+                models.TrialUserAttributeModel
+            ).values(trial_id=trial_id, key=key, value_json=json.dumps(value))
+            sqlite_upsert_stmt = sqlite_insert_stmt.on_conflict_do_update(
```

Comment:
`on_conflict_do_update` requires SQLite 3.24.0 or after. So, could you update the elif statement as follows?
```python
elif self.engine.name == "sqlite" and sqlite3.sqlite_version_info >= (3, 24, 0):
```

Reply:
Note: https://www.sqlite.org/changes.html#version_3_24_0